batches: update docs for publish from preview

courier-new opened the pull request:
Closes #23384 and #22179.

This PR:
- Updates the quick start doc to guide users through publishing their changesets from the preview UI, instead of with the batch spec file
- Adds steps to the "publishing changesets" doc to describe the process of publishing from the preview UI
- Adds an FAQ entry for why a changeset may be disabled from being selected in the preview UI
- Updates screenshots across quickstart + "publishing changesets"
- Adds a changelog entry

Since it's hard for you to see what the new docs or screenshots look like without pulling this and running it yourself, I took super long screenshots again to save you the trouble. 🙂

## Quickstart

![screencapture-localhost-5080-batch-changes-quickstart-2021-08-18-22_32_38](https://user-images.githubusercontent.com/8942601/130013706-1f702832-89d9-4a2a-a75d-b51205c8ba86.png)

## Publishing changesets

![screencapture-localhost-5080-batch-changes-how-tos-publishing-changesets-2021-08-18-22_33_29](https://user-images.githubusercontent.com/8942601/130013729-db66342e-0125-48d4-80ab-6950d5fefce3.png)

LawnGnome commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Code Insights will do a one-time reset of ephemeral insights specific database tables to clean up stale and invalid data. Insight data will regenerate automatically. [23791](https://github.com/sourcegraph/sourcegraph/pull/23791)
 - Perforce: added basic support for Perforce permission table path wildcards. [#23755](https://github.com/sourcegraph/sourcegraph/pull/23755)
 - Added autocompletion and search filtering of branch/tag/commit revisions to the repository compare page. [#23977](https://github.com/sourcegraph/sourcegraph/pull/23977)
+- Batch Changes changesets can now be [published from the Sourcegraph preview UI](https://docs.sourcegraph.com/batch_changes/how-tos/publishing_changesets#within-the-ui). [#22912](https://github.com/sourcegraph/sourcegraph/issues/22912)
```
I don't think "the Sourcegraph preview UI" will have meaning for users. How about:

```suggestion
- Batch Changes changesets can now be [set to published when previewing new or updated batch changes](https://docs.sourcegraph.com/batch_changes/how-tos/publishing_changesets#within-the-ui). [#22912](https://github.com/sourcegraph/sourcegraph/issues/22912)
```

LawnGnome commented on doc/batch_changes/how-tos/publishing_changesets.md:
```diff
@@ -85,7 +85,7 @@ A brief summary of the pros and cons of each workflow is:
           </li>
           <li>
-            Easy to publish random changesets without having to specify rules in the `published` field
+            Easy to publish random changesets without having to specify rules in the <code>published</code> field
```
This is obviously a styling nuance in docsite I was unaware of!

courier-new replied:
I was unaware as well, but I've seen similar nuances in other markdown environments when something is nested in a table. Guess this is one of those things. 😅

LawnGnome commented on doc/batch_changes/how-tos/publishing_changesets.md:
```diff
@@ -173,7 +173,19 @@ See [`changesetTemplate.published`](../references/batch_spec_yaml_reference.md#c
+
+#### From the preview
+
+When you run `src batch preview` against your batch spec and open the preview link, you'll see the current states of each of your changesets as well as the actions that will be performed when you apply:
```
Let's add a note here about the minimum version:

```suggestion
>NOTE: This feature was added in Sourcegraph 3.31.

When you run `src batch preview` against your batch spec and open the preview link, you'll see the current states of each of your changesets as well as the actions that will be performed when you apply:
```

courier-new replied:
Good callout, thanks!

LawnGnome commented on doc/batch_changes/how-tos/publishing_changesets.md:
```diff
@@ -173,7 +173,19 @@ See [`changesetTemplate.published`](../references/batch_spec_yaml_reference.md#c
+For any changesets that are currently unpublished or only published as drafts, you can select the checkbox and choose an action from the dropdown menu to indicate what publication state you want to set the changesets in on apply:
+<!-- TODO: Add image -->
+
+Once the preview actions look good, you can click "Apply" to publish the changesets.
```
Our [house style is to use bold for UI actions](https://about.sourcegraph.com/handbook/communication/content_guidelines/style_and_mechanics#instructions-references-and-citations), so:

```suggestion
Once the preview actions look good, you can click **Apply** to publish the changesets.
```

courier-new replied:
Ah, thank you! I even saw that while updating the other docs that mention this button, and it completely did not register. 🤦‍♀️

LawnGnome commented on doc/batch_changes/quickstart.md:
```diff
@@ -79,19 +78,16 @@ Let's see the changes that will be made. Don't worry---no commits, branches, or
-1. Click the **Apply spec** button to create the batch change. You should see something like this:
-    <img src="https://sourcegraphstatic.com/docs/images/batch_changes/browser_batch_created.png" class="screenshot">
-
-You created your first batch change! The batch change's changesets are still unpublished, which means they exist only on Sourcegraph and haven't been pushed to your code host yet.
```
I'm on the fence about removing this: I know that it means that we can't use the preview UI to publish the changesets in later steps if we remove it, but it leads into the explanation of publication state and batch change state being separate, which I think is valuable.

(More thoughts in a moment.)

LawnGnome commented on doc/batch_changes/quickstart.md:
```diff
@@ -109,31 +105,22 @@ This is a one-time operation that you don't need to do for each batch change. Yo
+### Publish to code hosts
 
-    > NOTE: You can also create or update a batch change by running `src batch apply`. This skips the preview stage, and is especially useful when updating an existing batch change.
+Back from the **preview page**, for each changest, you can choose whether to leave it as-is unpublished, publish it as a normal pull request/merge request, or publish it as a draft (if the code host supports it). Let's try publishing some changesets normally:
```
"Back from the preview page" doesn't really read well to me. If we do remove the **Apply** step above, then I would frame this as "Still on the preview page...".

That said, I might be more inclined to document this around the bulk action and link out to the new preview documentation as a "this is another way of doing this", rather than complicating the flow in this document.

It's also worth bearing in mind that we've just told the user on line 98 that they probably don't want to publish their changesets in the hello world example, and I'm not sure this really flows from that.

Thoughts?

courier-new replied:
> "Back from the preview page" doesn't really read well to me. If we do remove the Apply step above, then I would frame this as "Still on the preview page...".

Good point, I'll reword that based on what we decide to do with the previous section.

> That said, I might be more inclined to document this around the bulk action and link out to the new preview documentation as a "this is another way of doing this", rather than complicating the flow in this document.

(Followed up in main comment thread)

> It's also worth bearing in mind that we've just told the user on line 98 that they probably don't want to publish their changesets in the hello world example, and I'm not sure this really flows from that.

I've wondered about this since I onboarded as well, actually. 😅 We have a note in an earlier section that suggests a way to modify the spec file to something you'd actually want to test publishing with:

```
>NOTE: If you want to run the batch change on fewer repositories, change the `repositoriesMatchingQuery` in `hello-world.batch.yaml` to something like `file:README.md repo:myproject` (to only match repositories whose name contains `myproject`).
```

What if I move this out of a note and convert it to a first-class suggestion at line 98 where we point that out?

LawnGnome commented on doc/dev/background-information/batch_changes/index.md:
```diff
@@ -46,13 +46,13 @@ It starts in [`src-cli`](https://github.com/sourcegraph/src-cli):
 
 1. The [`BatchChangePreviewPage` component](https://github.com/sourcegraph/sourcegraph/blob/e7f26c0d7bc965892669a5fc9835ec65211943aa/client/web/src/enterprise/batches/preview/BatchChangePreviewPage.tsx#L43) then sends a GraphQL request to the backend to [query the `BatchSpecByID`](https://github.com/sourcegraph/sourcegraph/blob/e7f26c0d7bc965892669a5fc9835ec65211943aa/client/web/src/enterprise/ba ...
-1. Once you hit the "Apply spec" button, the component [uses the `applyBatchChange`](https://github.com/sourcegraph/sourcegraph/blob/e7f26c0d7bc965892669a5fc9835ec65211943aa/client/web/src/enterprise/batches/preview/backend.ts#L140-L159) to apply the batch spec and create a batch change.
+1. Once you hit the "Apply" button, the component [uses the `applyBatchChange`](https://github.com/sourcegraph/sourcegraph/blob/e7f26c0d7bc965892669a5fc9835ec65211943aa/client/web/src/enterprise/batches/preview/backend.ts#L140-L159) to apply the batch spec and create a batch change.
```
I know it was wrong before, but let's fix this while we're here:

```suggestion
1. Once you hit the **Apply** button, the component [uses the `applyBatchChange`](https://github.com/sourcegraph/sourcegraph/blob/e7f26c0d7bc965892669a5fc9835ec65211943aa/client/web/src/enterprise/batches/preview/backend.ts#L140-L159) to apply the batch spec and create a batch change.
```

We should also update that link.

courier-new replied:
Hmm, strangely, it looks like a lot of links on this page refer to a specific commit hash. It also looks like the line numbering is off on a number of them when I try to update. 😬 Mind if I leave this off for now, and we try to update them all together?

LawnGnome replied:
Sure thing. I would expect the line numbers will have to change, but yeah, we can handle these en masse.

LawnGnome commented on doc/dev/background-information/batch_changes/index.md:
```diff
@@ -46,13 +46,13 @@ It starts in [`src-cli`](https://github.com/sourcegraph/src-cli):
 
 1. The [`CreateChangesetSpec`](https://github.com/sourcegraph/sourcegraph/blob/8b99439e21aaa000443382f03f92e532b0445858/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go#L545) and [`CreateBatchSpec`](https://github.com/sourcegraph/sourcegraph/blob/8b99439e21aaa000443382f03f92e532b0445858/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go#L489) mutations that `src-cli` cal ...
-1. When you clicked "Apply Spec" the [`ApplyBatchChange` resolver](https://github.com/sourcegraph/sourcegraph/blob/8b99439e21aaa000443382f03f92e532b0445858/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go#L404) was executed to create the batch change.
+1. When you clicked "Apply" the [`ApplyBatchChange` resolver](https://github.com/sourcegraph/sourcegraph/blob/8b99439e21aaa000443382f03f92e532b0445858/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go#L404) was executed to create the batch change.
```
```suggestion
1. When you clicked **Apply** the [`ApplyBatchChange` resolver](https://github.com/sourcegraph/sourcegraph/blob/8b99439e21aaa000443382f03f92e532b0445858/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go#L404) was executed to create the batch change.
```

And again, this link should be recreated against our current `main`.

LawnGnome reviewed (commented):
This generally LGTM, but I wonder if I wrote the ticket here too narrowly: the quickstart changes are simultaneously what I think I scoped out in that ticket, and probably inappropriate in context right now. (Totally not your fault! You wrote exactly what I thought we'd write; I just thought wrong.)

Beyond the quick start changes, though, I think this is solid. In the future, I'd probably prefer to see the documentation and non-documentation changes separated out, but I don't think it matters here.

courier-new commented:
> This generally LGTM, but I wonder if I wrote the ticket here too narrowly: the quickstart changes are simultaneously what I think I scoped out in that ticket, and probably inappropriate in context right now. (Totally not your fault! You wrote exactly what I thought we'd write; I just thought wrong.)

Bummer, yeah, I see what you mean with the comments you left... I'm honestly not sure what the "easiest" way to publish changesets is right now, and I feel like I need to have an answer to that to be able to recommend what steps we include on the quick start... My gut still leans ever-so-slightly towards recommending the preview way, though, and I think the main reason is that, to get from zero to published changesets:

- it's fewer steps in the guide
- it's fewer distinct GUIs/TUIs to take in (terminal + preview vs. terminal + preview + details)
- it's less overall information to digest upfront (I agree the publication state vs. batch change state distinction is valuable, but maybe not critical to understand as a first time user?)

But I'm not sure, I'm probably extremely biased towards this direction since I took over this part of the implementation from you and already wrote something. 😛

> Beyond the quick start changes, though, I think this is solid. In the future, I'd probably prefer to see the documentation and non-documentation changes separated out, but I don't think it matters here.

Right, absolutely. That's just laziness on my part because I wanted to wrap this up sooner. 😅

LawnGnome commented:
> I see what you mean with the comments you left... I'm honestly not sure what the "easiest" way to publish changesets is right now, and I feel like I need to have an answer to that to be able to recommend what steps we include on the quick start.

I think that's a good insight and question. Maybe something we should discuss at sync tomorrow to see what others think?

courier-new commented:
I do believe that would be a great use of the sync! I'll add that promptly. 🙂

courier-new commented:
@eseliger I just remembered I forgot to add the FAQ entry you suggested; I'll add that tomorrow morning before merging this.

eseliger commented on doc/batch_changes/how-tos/publishing_changesets.md:
```diff
@@ -173,7 +173,21 @@ See [`changesetTemplate.published`](../references/batch_spec_yaml_reference.md#c
+> NOTE: This feature requires Sourcegraph 3.31 or later.
+
+When you run `src batch preview` against your batch spec and open the preview link, you'll see the current states of each of your changesets as well as the actions that will be performed when you apply:
+<!-- TODO: Add image -->
```
reminder :)

eseliger reviewed (approved):
Mirroring what Adam said, otherwise LGTM !

courier-new commented:
Following conversation in team sync this morning, going forward we will make two changes to this approach:

1) Make the recommended way from the quick start guide to apply the batch change, then publish changesets with bulk actions
2) Formally suggest revising and reapplying the spec with a single repo targeted in order to be able to publish a changeset while going through the quick start

courier-new commented:
Rebased on `main`, updated the quickstart guide, and replaced all the relevant screenshots! Please take another look (I added screenshots of the new docs themselves to the PR description for your perusal!) and let me know if you have any additional feedback. 🙂

eseliger commented on doc/batch_changes/references/faq.md:
```diff
@@ -72,3 +71,14 @@ Keep in mind the context in which the inner `${{ }}` will be evaluated and be su
+
+Since Sourcegraph 3.31, it is possible to publish many types of changeset when previewing a batch change by modifying the publication state for the changeset directly from the UI (see ["Publishing changesets"](../how-tos/publishing_changesets.md#from-the-preview)). However, not every changeset can be published by Sourcegraph. By hovering over your changeset's disabled checkbox, you can see the re ...
+
+- The changeset is already published (we cannot unpublish a changeset, or convert it back to a draft).
```
Back to draft should be possible, I think 🤔

courier-new replied:
I asked Adam about this while working on #24067 and he pointed me to #17541, which mentions `true => draft` still isn't possible. I also didn't see a test for this transition on `plan_test.go` with all the others. But maybe you know something more? 👀

LawnGnome reviewed (approved):
This is great! The flow on this is much better now, particularly in the quick start.

I've made a few wording suggestions, but don't feel like you need to take any of them: they're just places where I think we can be more precise (but maybe at the expense of readability). Otherwise, :+1:.
